crates/ruff_formatter/src/format_element.rs

```diff
@@ -284,19 +301,32 @@ impl FormatElements for FormatElement {
     }
 }
 
-/// Provides the printer with different representations for the same element so that the printer
-/// can pick the best fitting variant.
-///
-/// Best fitting is defined as the variant that takes the most horizontal space but fits on the line.
-#[derive(Clone, Eq, PartialEq)]
-pub struct BestFitting {
-    /// The different variants for this element.
-    /// The first element is the one that takes up the most space horizontally (the most flat),
-    /// The last element takes up the least space horizontally (but most horizontal space).
-    variants: Box<[Box<[FormatElement]>]>,
+/// Mode used to determine if any variant (except the most expanded) fits for [`BestFittingVariants`].
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Default)]
+pub enum BestFittingMode {
+    /// The variant fits if the content up to the first [`hard_line_break`] or a soft line break inside a [`Group`] with
+    /// [`PrintMode::Expand`] fits on the line. The default mode.
+    #[default]
+    FirstLine,
+
+    /// A variant fits if all lines fit into the configured print width. A line ends if by any
+    /// [`hard_line_break`] or a soft line break inside a [`Group`] with [`PrintMode::Expand`].
+    /// The content doesn't fit if there's any hard line break  outside a [`Group`] with [`PrintMode::Expand`]
+    /// (a hard line break in content that should be considered in [`PrintMode::Flat`].
+    ///
+    /// Use this mode with caution as it requires measuring all content of the variant which is more
+    /// expensive than using [`BestFittingMode::FirstLine`].
+    AllLines,
 }
 
-impl BestFitting {
+/// The different variants for this element.
+/// The first element is the one that takes up the most space horizontally (the most flat),
+/// The last element takes up the least space horizontally (but most horizontal space).
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct BestFittingVariants(Box<[Box<[FormatElement]>]>);
```

Comment:
I couldn't add `mode` here because it would increase the size of the struct to 24 bytes which, in turn, would increase the size of `FormatElement` from 24 to 32 bytes. Adding `mode` inline avoids the size increase